### PR TITLE
File-upload: smooth time remaining, bitrate and stabilize user information

### PR DESCRIPTION
**Pull request by Phlogi**
Improving the upload progress bar's "Time remaining".
Tested with a single large file and multiple smaller files. 

Besides the existing moving average, a smoothing factor is introduced for the time remaining display as well as the bitrate. 
Furthermore, half of the buffer needs to be filled before the first prediction is displayed to the user. This reduces volatile and jumping durations towards the user and improves usability.

### CODE REVIEW

File: apps/files/js/file-upload.js [hunk 1]
```diff
@@ -1192,33 +1197,54 @@ OC.Uploader.prototype = _.extend({
...
+
+					if (smoothRemainingSeconds == null){
 						smoothRemainingSeconds = bufferTotal / bufferSize;
-					} else {
-						smoothRemainingSeconds = 1;
+					} else{
+						smoothRemainingSeconds = smoothing * (bufferTotal / bufferSize) + ((1-smoothing) * smoothRemainingSeconds);
+					}
+
+					if (bufferIndex % 4 == 0){
```
Comment (CarlSchwan) [inline apps/files/js/file-upload.js:1227]:
```suggestion
					if (bufferIndex % 4 === 0) {
```

File: apps/files/js/file-upload.js [hunk 2]
```diff
@@ -1192,33 +1197,54 @@ OC.Uploader.prototype = _.extend({
...
-					var h = moment.duration(smoothRemainingSeconds, "seconds").humanize();
-					if (!(smoothRemainingSeconds >= 0 && smoothRemainingSeconds < 14400)) {
+					// wait for the buffer to be at least half filled, approximately takes 3s
+					if (!(smoothRemainingSeconds >= 0 && smoothRemainingSeconds < 14400) || (bufferFilled == false && bufferIndex < bufferSize/2)) {
 						// show "Uploading ..." for durations longer than 4 hours
 						h = t('files', 'Uploading …');
 					}
+
+					// smooth bitrate
+					if (smoothBitrate == null){
```
Comment (CarlSchwan) [inline apps/files/js/file-upload.js:1238]:
```suggestion
					if (smoothBitrate === null) {
```

File: apps/files/js/file-upload.js [hunk 3]
```diff
@@ -1192,33 +1197,54 @@ OC.Uploader.prototype = _.extend({
...
 						buffer[bufferIndex] = remainingSeconds; //buffer to make it smoother
+
 						bufferIndex = (bufferIndex + 1) % bufferSize;
-						bufferIndex2++;
 					}
-					var smoothRemainingSeconds;
-					if (bufferIndex2 > 0 && bufferIndex2 < 20) {
-						smoothRemainingSeconds = bufferTotal / bufferIndex2;
-					} else if (bufferSize > 0) {
+					if (bufferIndex == bufferSize - 1){
```
Comment (CarlSchwan) [inline apps/files/js/file-upload.js:1216]:
```suggestion
					if (bufferIndex === bufferSize - 1) {
```

File: apps/files/js/file-upload.js [hunk 4]
```diff
@@ -1192,33 +1197,54 @@ OC.Uploader.prototype = _.extend({
...
+						bufferFilled = true;
+					}
+					//console.log('#', ' idx: ',bufferIndex, ' Total: ', bufferTotal, ' remainSeconds: ', remainingSeconds, ' during: ', diffUpdate);
+
+					if (smoothRemainingSeconds == null){
```
Comment (CarlSchwan) [inline apps/files/js/file-upload.js:1221]:
```suggestion
					if (smoothRemainingSeconds === null) {
```

File: apps/files/js/file-upload.js [hunk 5]
```diff
@@ -1192,33 +1197,54 @@ OC.Uploader.prototype = _.extend({
 					var diffUpdate = (thisUpdate - lastUpdate)/1000; // eg. 2s
 					lastUpdate = thisUpdate;
 					var diffSize = data.loaded - lastSize;
+					if (diffSize <= 0){
```
Comment (CarlSchwan) [inline apps/files/js/file-upload.js:1200]:
```suggestion
					if (diffSize <= 0) {
```

File: apps/files/js/file-upload.js [hunk 6]
```diff
@@ -1151,7 +1152,7 @@ OC.Uploader.prototype = _.extend({
 
 			if (this._supportAjaxUploadWithProgress()) {
 				//remaining time
-				var lastUpdate, lastSize, bufferSize, buffer, bufferIndex, bufferIndex2, bufferTotal;
+				var lastUpdate, lastSize, bufferSize, buffer, bufferIndex, bufferTotal, smoothRemainingSeconds, smoothBitrate;
```
Comment (CarlSchwan) [inline apps/files/js/file-upload.js:1155]: Ideally declare the new variables near where you are using them. This makes the code easier to read and not look like very old-style C code ;)

**Review by CarlSchwan — COMMENTED** [review]
Very cool feature, I left some code style but I didn't test it yet locally. I will give it a try tomorrow :)

File: apps/files/js/file-upload.js (hunk 6 above)
Comment (Phlogi) [reply apps/files/js/file-upload.js:1155]: Well, these need to persist over calling the individual functions using them, so I put them here (not 100% sure if that is necessary). I might have just followed the existing place in the source.

**Comment by Phlogi** [comment]
Any update?

**Review by szaimen — CHANGES_REQUESTED** [review]
I just tested this and somehow the speed does not seem to work correctly:
image

**Review by artonge — CHANGES_REQUESTED** [review]
Also, canceling an upload seems to fail now

File: apps/files/js/file-upload.js [hunk 7]
```diff
@@ -1192,33 +1197,54 @@ OC.Uploader.prototype = _.extend({
...
+						bufferFilled = true;
+					}
+					//console.log('#', ' idx: ',bufferIndex, ' Total: ', bufferTotal, ' remainSeconds: ', remainingSeconds, ' during: ', diffUpdate);
+
+					if (smoothRemainingSeconds === null) {
 						smoothRemainingSeconds = bufferTotal / bufferSize;
-					} else {
-						smoothRemainingSeconds = 1;
+					} else{
+						smoothRemainingSeconds = smoothing * (bufferTotal / bufferSize) + ((1-smoothing) * smoothRemainingSeconds);
```
[the remark this answers is not in the document]
Comment (Phlogi) [reply apps/files/js/file-upload.js:1224]: It should check for undefined. 
`					if (smoothRemainingSeconds === undefined){`

Same for the bitrate:

`					if (smoothBitrate === undefined){`

I just tested this and the `NaN` problem is fixed.

**Comment by Phlogi** [comment]
> Also, canceling an upload seems to fail now

I can reproduce this, however I don't think it's related to the change here. I see a 404 in the frontend:
image


Here is the message I see in the debug log on the server side: 

```
{
  "reqId": "QCf3mPDlqWTxKByNmFtN",
  "level": 3,
  "time": "2022-04-18T19:24:22+00:00",
  "remoteAddr": "192.168.1.244",
  "user": "webadmin",
  "app": "core",
  "method": "PUT",
  "url": "/remote.php/dav/uploads/webadmin/web-file-upload-5368d4c4d0bf8c192c21c77fb3e39e95-1650309856915/0",
  "message": "unable to rename, source directory is not writable : uploads/web-file-upload-5368d4c4d0bf8c192c21c77fb3e39e95-1650309856915",
  "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:99.0) Gecko/20100101 Firefox/99.0",
  "version": "22.2.3.0"
}
{
  "reqId": "QCf3mPDlqWTxKByNmFtN",
  "level": 3,
  "time": "2022-04-18T19:24:22+00:00",
  "remoteAddr": "192.168.1.244",
  "user": "webadmin",
  "app": "webdav",
  "method": "PUT",
  "url": "/remote.php/dav/uploads/webadmin/web-file-upload-5368d4c4d0bf8c192c21c77fb3e39e95-1650309856915/0",
  "message": "renaming part file to final file failed $renameOkay: false, $fileExists: false)",
  "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:99.0) Gecko/20100101 Firefox/99.0",
  "version": "22.2.3.0"
}
{
  "reqId": "QCf3mPDlqWTxKByNmFtN",
  "level": 4,
  "time": "2022-04-18T19:24:22+00:00",
  "remoteAddr": "192.168.1.244",
  "user": "webadmin",
  "app": "webdav",
  "method": "PUT",
  "url": "/remote.php/dav/uploads/webadmin/web-file-upload-5368d4c4d0bf8c192c21c77fb3e39e95-1650309856915/0",
  "message": "Could not rename part file to final file",
  "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:99.0) Gecko/20100101 Firefox/99.0",
  "version": "22.2.3.0",
  "exception": {
    "Exception": "Sabre\\DAV\\Exception",
    "Message": "Could not rename part file to final file",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php",
        "line": 155,
        "function": "put",
        "class": "OCA\\DAV\\Connector\\Sabre\\File",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/dav/lib/Upload/UploadFolder.php",
        "line": 45,
        "function": "createFile",
        "class": "OCA\\DAV\\Connector\\Sabre\\Directory",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
        "line": 1098,
        "function": "createFile",
        "class": "OCA\\DAV\\Upload\\UploadFolder",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php",
        "line": 504,
        "function": "createFile",
        "class": "Sabre\\DAV\\Server",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php",
        "line": 89,
        "function": "httpPut",
        "class": "Sabre\\DAV\\CorePlugin",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
        "line": 472,
        "function": "emit",
        "class": "Sabre\\DAV\\Server",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
        "line": 253,
        "function": "invokeMethod",
        "class": "Sabre\\DAV\\Server",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php",
        "line": 321,
        "function": "start",
        "class": "Sabre\\DAV\\Server",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/dav/lib/Server.php",
        "line": 333,
        "function": "exec",
        "class": "Sabre\\DAV\\Server",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/apps/dav/appinfo/v2/remote.php",
        "line": 35,
        "function": "exec",
        "class": "OCA\\DAV\\Server",
        "type": "->"
      },
      {
        "file": "/var/www/nextcloud/remote.php",
        "line": 166,
        "args": [
          "/var/www/nextcloud/apps/dav/appinfo/v2/remote.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/nextcloud/apps/dav/lib/Connector/Sabre/File.php",
    "Line": 295,
    "CustomMessage": "--"
  }
}
```

**Comment by szaimen** [comment]
> Also, canceling an upload seems to fail now

@artonge this is unrelated, afaik.
I think it is failing in master and stable24 too...

**Comment by szaimen** [comment]
Interesting approach to merge it into a different branch in our org :)

**Pull request merged**